Creates/Updates stack with parameters to template files

robertstettner opened the pull request:
This adds the ability to use the AWS CloudFormation in-built parameters for a template file.

I added tests for it too.

mvillis commented:
Was just looking for this! Nice one.
Would be good to see the readme updated :-)

andyday commented:
this was on the roadmap.  great addition.  code looks good.  need to add documentation to the readme.

andyday commented on index.js:
```diff
@@ -271,7 +281,8 @@ function Cfn(name, template) {
                     Capabilities: capabilities,
-                    TemplateBody: data
+                    TemplateBody: data,
+                    Parameters: convertParams(params)
```
params here refers to function template parameters.  going forward it should server 2 purposes.

1. If the template is a js function template, this object should be passed into the function
2. If it is any other template it can be converted as passed in as CloudFormation parameters.

andyday commented on test/index.js:
```diff
@@ -126,6 +126,29 @@ describe('create/update', function() {
                         updateStackStub.stub.should.be.calledOnce();
                     });
             });
+            it('updates stack with parameters', function(){
```
The fact that a test for function templates with parameters, did not fail is concerning...  Not absolutely necessary, but it would be nice to add this test...

robertstettner commented:
I will get around to doing this. I will be on holiday for most of April, and I simply cannot do it right now as I'm busy at work.

robertstettner commented:
Finally got around to doing this. Added a few things.

- updated the documentation as requested
- updated logic around JS-module templates and not passing `Parameters` into the stack options
- added more tests as requested
- mentioned YAML format in the docs
- added the ability to use YAML string templates
- added the ability to use serialized JSON string templates

andyday commented on index.js:
```diff
@@ -247,32 +248,71 @@ function Cfn(name, template) {
             // Check if template if a `js` file
             case _.endsWith(template, '.js'):
                 promise = loadJs(template);
+                delete stackOptions.Parameters;
```
we should only delete parameters if the .js template is a module function.  there could be the case where a .js uses traditional cfn parameters.

robertstettner replied:
These are not the parameters passed in the option. They are the Cloudformation generated ones from the original options. I believe the `loadJs` interpolates them into the JS-module anyways, and the actual call to the AWS Cloudformation SDK doesn't need the `Parameters` in that case.

robertstettner replied:
Hang on, should there really be that case? Because... 

In the case of using this with a JS-module that uses both the Cloudformation parameters and one's that are interpolated in, the parameters as options *should be decoupled*.
For example:
```js
cfn({
  name: 'foo',
  template: 'template.js',
  params: { ... }, // only interpolated JS-module parameters
  cfParams: { ... } // only Cloudformation parameters
});
```

andyday replied:
I think I like this more explicit approach.  It clears up any potential confusion...

andyday replied:
Let's go with that...

robertstettner commented:
I think this PR is ready to go, unless there is anything else?

andyday commented:
published as 1.0.1
